File: unity-package/Editor/Generation/BatchExpansion.cs
using System;
using System.Collections.Generic;

namespace UnityAiAssets.Editor.Generation
{
    public enum BatchSeedModeKind
    {
        Fixed = 0,
        Random = 1,
        Sequential = 2
    }

    public sealed class BatchExpansionItem
    {
        public int Index;
        public int PromptIndex;
        public int VariationIndex;
        public long Seed;
        public string Prompt;
        public string OutputName;
    }

    public sealed class BatchExpansionPlan
    {
        public readonly List<BatchExpansionItem> Items = new List<BatchExpansionItem>();
        public string SeedMode = "fixed";
        public readonly List<long> BaseSeeds = new List<long>();
        public int VariationCount = 1;
        public readonly List<string> Warnings = new List<string>();

        public int JobCount => Items.Count;

        public List<long> SeedsForPrompt(int promptIndex = 0)
        {
            var seeds = new List<long>();
            foreach (var item in Items)
            {
                if (item.PromptIndex == promptIndex)
                    seeds.Add(item.Seed);
            }

            return seeds;
        }

        public string SeedSummary()
        {
            var seeds = SeedsForPrompt(0);
            if (seeds.Count == 0)
                return "No seeds";
            if (seeds.Count <= 12)
                return string.Join(", ", seeds);
            return seeds[0] + "…" + seeds[seeds.Count - 1] + " (" + seeds.Count + " unique seeds per prompt)";
        }
    }

    /// <summary>
    /// Deterministic prompt × seed × variation expansion. Must match
    /// <c>unity_ai_assets.services.batch_expansion</c>.
    /// </summary>
    public static class BatchExpansion
    {
        public const int DefaultMaxJobs = 32;
        public const int DefaultMaxPrompts = 50;
        public const int DefaultMaxVariations = 16;
        public const int WarnJobCount = 16;

        public static string ToApiValue(BatchSeedModeKind mode)
        {
            switch (mode)
            {
                case BatchSeedModeKind.Fixed: return "fixed";
                case BatchSeedModeKind.Sequential: return "sequential";
                default: return "random";
            }
        }

        public static string BuildOutputName(
            string baseName, int promptIndex, long seed, int variationIndex, int maxLength)
        {
            var stem = string.IsNullOrWhiteSpace(baseName) ? "texture" : baseName.Trim();
            var suffix = "_p" + promptIndex.ToString("00") + "_s" + seed + "_v" + variationIndex.ToString("00");
            if (suffix.Length >= maxLength)
            {
                var start = suffix.Length - maxLength;
                return suffix.Substring(start);
            }

            var room = maxLength - suffix.Length;
            var trimmed = stem.Length <= room ? stem : stem.Substring(0, room);
            trimmed = trimmed.TrimEnd('_', '-', '.');
            if (string.IsNullOrEmpty(trimmed))
                trimmed = "tex";
            return trimmed + suffix;
        }

        public static bool TryExpand(
            IList<string> prompts,
            BatchSeedModeKind seedMode,
            int variationCount,
            long seed,
            long seedStart,
            long seedEnd,
            string outputName,
            out BatchExpansionPlan plan,
            out List<string> errors,
            int maxJobs = DefaultMaxJobs,
            int maxPrompts = DefaultMaxPrompts,
            int maxVariations = DefaultMaxVariations,
            int maxPromptLength = 2000,
            int maxOutputNameLength = 100,
            long minSeed = 0,
            long maxSeed = 4294967295)
        {
            errors = new List<string>();
            plan = new BatchExpansionPlan { VariationCount = variationCount, SeedMode = ToApiValue(seedMode) };
            if (prompts == null || prompts.Count == 0)
            {
                errors.Add("At least one prompt is required.");
                return false;
            }

            if (prompts.Count > maxPrompts)
            {
                errors.Add("A batch may include at most " + maxPrompts + " prompts.");
                return false;
            }

            var normalized = new List<string>();
            for (var i = 0; i < prompts.Count; i++)
            {
                var text = NormalizePrompt(prompts[i]);
                if (string.IsNullOrEmpty(text))
                {
                    errors.Add("Prompt " + (i + 1) + " is empty.");
                    continue;
                }

                if (text.Length > maxPromptLength)
                {
                    errors.Add("Prompt " + (i + 1) + " exceeds the maximum length of " + maxPromptLength + ".");
                    continue;
                }

                normalized.Add(text);
            }

            if (errors.Count > 0)
                return false;

            if (variationCount < 1)
            {
                errors.Add("Variation count must be at least 1.");
                return false;
            }

            if (variationCount > maxVariations)
            {
                errors.Add("Variation count may be at most " + maxVariations + ".");
                return false;
            }

            if (!TryResolveBaseSeeds(
                    seedMode, seed, seedStart, seedEnd, minSeed, maxSeed,
                    plan.BaseSeeds, errors))
                return false;

            var stride = plan.BaseSeeds.Count;
            var jobCount = normalized.Count * stride * variationCount;
            if (jobCount > maxJobs)
            {
                errors.Add(
                    "This batch would create " + jobCount + " jobs, which exceeds the maximum of " +
                    maxJobs + ". Reduce prompts, seed range, or variations.");
                return false;
            }

            var seenPrompts = new HashSet<string>();
            for (var promptIndex = 0; promptIndex < normalized.Count; promptIndex++)
            {
                var prompt = normalized[promptIndex];
                if (!seenPrompts.Add(prompt))
                    plan.Warnings.Add(
                        "Prompt " + (promptIndex + 1) +
                        " duplicates an earlier entry and will reuse the same seed set.");

                var usedSeeds = new HashSet<long>();
                foreach (var baseSeed in plan.BaseSeeds)
                {
                    for (var variationIndex = 0; variationIndex < variationCount; variationIndex++)
                    {
                        var actual = baseSeed + variationIndex * (long)stride;
                        if (actual > maxSeed)
                        {
                            errors.Add(
                                "Expanded seed exceeds the allowed maximum. Reduce the sequential range or variation count.");
                            return false;
                        }

                        if (!usedSeeds.Add(actual))
                        {
                            errors.Add("Sequential seed expansion produced a duplicate seed for the same prompt.");
                            return false;
                        }

                        plan.Items.Add(new BatchExpansionItem
                        {
                            Index = plan.Items.Count,
                            PromptIndex = promptIndex,
                            VariationIndex = variationIndex,
                            Seed = actual,
                            Prompt = prompt,
                            OutputName = BuildOutputName(
                                outputName, promptIndex, actual, variationIndex, maxOutputNameLength)
                        });
                    }
                }
            }

            return true;
        }

        static bool TryResolveBaseSeeds(
            BatchSeedModeKind seedMode,
            long seed,
            long seedStart,
            long seedEnd,
            long minSeed,
            long maxSeed,
            List<long> baseSeeds,
            List<string> errors)
        {
            if (seedMode == BatchSeedModeKind.Sequential)
            {
                if (seedStart < minSeed || seedStart > maxSeed || seedEnd < minSeed || seedEnd > maxSeed)
                {
                    errors.Add("Sequential seed range must stay within " + minSeed + "…" + maxSeed + ".");
                    return false;
                }

                if (seedStart > seedEnd)
                {
                    errors.Add("Seed start must be less than or equal to seed end.");
                    return false;
                }

                for (var value = seedStart; value <= seedEnd; value++)
                    baseSeeds.Add(value);
                return true;
            }

            if (seed < minSeed || seed > maxSeed)
            {
                errors.Add("Seed must be between " + minSeed + " and " + maxSeed + ".");
                return false;
            }

            if (seedMode == BatchSeedModeKind.Fixed || seedMode == BatchSeedModeKind.Random)
            {
                baseSeeds.Add(seed);
                return true;
            }

            errors.Add("Unknown seed mode.");
            return false;
        }

        static string NormalizePrompt(string raw)
        {
            if (string.IsNullOrWhiteSpace(raw))
                return string.Empty;
            return string.Join(" ", raw.Split((char[])null, StringSplitOptions.RemoveEmptyEntries));
        }
    }
}

File: src/unity_ai_assets/services/batch_expansion.py
"""Deterministic expansion of batch prompts, seeds, and variations into job specs.

A batch does not run inference. It only produces an ordered list of ordinary
generation jobs with unique seeds per prompt when sequential variation is used.
"""

from __future__ import annotations

import secrets
from collections.abc import Callable
from dataclasses import dataclass

from unity_ai_assets.core.error_codes import FieldIssueCode
from unity_ai_assets.core.errors import (
    BatchRequestInvalidError,
    BatchTooLargeError,
    FieldIssue,
)
from unity_ai_assets.domain.enums import KNOWN_BATCH_SEED_MODES, BatchSeedMode


@dataclass(frozen=True, slots=True)
class BatchExpansionItem:
    """One expanded generation job specification."""

    index: int
    prompt_index: int
    variation_index: int
    seed: int
    prompt: str
    output_name: str


@dataclass(frozen=True, slots=True)
class BatchExpansionPlan:
    """Validated, ordered expansion used for preview and job creation."""

    items: tuple[BatchExpansionItem, ...]
    seed_mode: str
    base_seeds: tuple[int, ...]
    variation_count: int
    warnings: tuple[str, ...] = ()

    @property
    def job_count(self) -> int:
        return len(self.items)

    @property
    def prompt_count(self) -> int:
        return len({item.prompt_index for item in self.items})

    def seeds_for_prompt(self, prompt_index: int = 0) -> tuple[int, ...]:
        return tuple(item.seed for item in self.items if item.prompt_index == prompt_index)

    def seed_summary(self) -> str:
        seeds = self.seeds_for_prompt(0)
        if not seeds:
            return "No seeds"
        if len(seeds) <= 12:
            return ", ".join(str(seed) for seed in seeds)
        return f"{seeds[0]}…{seeds[-1]} ({len(seeds)} unique seeds per prompt)"


def build_output_name(
    base: str,
    *,
    prompt_index: int,
    seed: int,
    variation_index: int,
    max_length: int,
) -> str:
    """Build a predictable, collision-safe output stem for one expanded job."""
    stem = (base or "texture").strip() or "texture"
    suffix = f"_p{prompt_index:02d}_s{seed}_v{variation_index:02d}"
    if len(suffix) >= max_length:
        return suffix[-max_length:]
    room = max_length - len(suffix)
    trimmed = stem[:room].rstrip("_-.")
    if not trimmed:
        trimmed = "tex"
        if len(trimmed) + len(suffix) > max_length:
            return suffix[-max_length:]
    return trimmed + suffix


def _issue(message: str, *, actual: object = None) -> FieldIssue:
    return FieldIssue(code=FieldIssueCode.VALUE_INVALID, message=message, actual=actual)


def _required(message: str) -> FieldIssue:
    return FieldIssue(code=FieldIssueCode.FIELD_REQUIRED, message=message)


def normalize_prompts(
    prompts: list[str],
    *,
    max_prompts: int,
    max_prompt_length: int,
) -> list[str]:
    """Strip and validate prompt entries, preserving caller order."""
    if not prompts:
        raise BatchRequestInvalidError(
            "At least one prompt is required.",
            field_issues={"prompts": [_required("At least one prompt is required.")]},
        )
    if len(prompts) > max_prompts:
        raise BatchRequestInvalidError(
            f"A batch may include at most {max_prompts} prompts.",
            field_issues={
                "prompts": [
                    FieldIssue(
                        code=FieldIssueCode.VALUE_ABOVE_MAXIMUM,
                        message=f"A batch may include at most {max_prompts} prompts.",
                        actual=len(prompts),
                        maximum=max_prompts,
                    )
                ]
            },
        )
    normalized: list[str] = []
    issues: dict[str, list[FieldIssue]] = {}
    for index, raw in enumerate(prompts):
        text = " ".join(str(raw or "").split())
        field = f"prompts.{index}"
        if not text:
            issues[field] = [_required("Prompt must not be empty.")]
            continue
        if len(text) > max_prompt_length:
            issues[field] = [
                FieldIssue(
                    code=FieldIssueCode.VALUE_TOO_LONG,
                    message=f"Prompt exceeds the maximum length of {max_prompt_length}.",
                    actual=len(text),
                    maximum=max_prompt_length,
                )
            ]
            continue
        normalized.append(text)
    if issues:
        raise BatchRequestInvalidError(
            "One or more prompt entries are invalid.",
            field_issues=issues,
        )
    return normalized


def resolve_base_seeds(
    *,
    seed_mode: str,
    seed: int | None,
    seed_start: int | None,
    seed_end: int | None,
    variation_count: int,
    min_seed: int,
    max_seed: int,
    random_seed_factory: Callable[[int, int], int] | None = None,
) -> tuple[int, ...]:
    """Resolve the base seed list used before variation offsets."""
    mode = (seed_mode or "").strip().lower()
    if mode not in KNOWN_BATCH_SEED_MODES:
        raise BatchRequestInvalidError(
            "seed_mode must be fixed, random, or sequential.",
            field_issues={
                "seed_mode": [
                    _issue(
                        "seed_mode must be fixed, random, or sequential.",
                        actual=seed_mode,
                    )
                ]
            },
        )
    if variation_count < 1:
        raise BatchRequestInvalidError(
            "variation_count must be at least 1.",
            field_issues={
                "variation_count": [
                    FieldIssue(
                        code=FieldIssueCode.VALUE_BELOW_MINIMUM,
                        message="variation_count must be at least 1.",
                        actual=variation_count,
                        minimum=1,
                    )
                ]
            },
        )

    def _in_range(value: int, field_name: str) -> None:
        if value < min_seed or value > max_seed:
            raise BatchRequestInvalidError(
                f"{field_name} is outside the allowed seed range.",
                field_issues={
                    field_name: [
                        FieldIssue(
                            code=FieldIssueCode.VALUE_INVALID,
                            message=f"{field_name} must be between {min_seed} and {max_seed}.",
                            actual=value,
                            minimum=min_seed,
                            maximum=max_seed,
                        )
                    ]
                },
            )

    if mode == BatchSeedMode.SEQUENTIAL.value:
        if seed_start is None or seed_end is None:
            raise BatchRequestInvalidError(
                "Sequential seed mode requires seed_start and seed_end.",
                field_issues={
                    "seed_start": [_required("seed_start is required for sequential mode.")],
                    "seed_end": [_required("seed_end is required for sequential mode.")],
                },
            )
        start = int(seed_start)
        end = int(seed_end)
        _in_range(start, "seed_start")
        _in_range(end, "seed_end")
        if start > end:
            raise BatchRequestInvalidError(
                "seed_start must be less than or equal to seed_end.",
                field_issues={
                    "seed_start": [
                        _issue("seed_start must be less than or equal to seed_end.", actual=start)
                    ]
                },
            )
        return tuple(range(start, end + 1))

    if mode == BatchSeedMode.FIXED.value:
        if seed is None:
            raise BatchRequestInvalidError(
                "Fixed seed mode requires a seed.",
                field_issues={"seed": [_required("seed is required for fixed mode.")]},
            )
        resolved = int(seed)
        _in_range(resolved, "seed")
        return (resolved,)

    if seed is not None:
        resolved = int(seed)
        _in_range(resolved, "seed")
        return (resolved,)
    factory = random_seed_factory or (
        lambda low, high: secrets.randbelow(high - low + 1) + low
    )
    span = max(0, variation_count - 1)
    high = max_seed - span
    if high < min_seed:
        raise BatchRequestInvalidError(
            "variation_count is too large to fit in the allowed seed range.",
            field_issues={
                "variation_count": [
                    _issue(
                        "variation_count is too large to fit in the allowed seed range.",
                        actual=variation_count,
                    )
                ]
            },
        )
    return (int(factory(min_seed, high)),)


def expand_batch(
    prompts: list[str],
    *,
    seed_mode: str,
    variation_count: int,
    seed: int | None = None,
    seed_start: int | None = None,
    seed_end: int | None = None,
    output_name: str = "texture",
    min_seed: int = 0,
    max_seed: int = 2**32 - 1,
    max_jobs: int = 32,
    max_prompts: int = 50,
    max_variations: int = 16,
    max_prompt_length: int = 2000,
    max_output_name_length: int = 100,
    random_seed_factory: Callable[[int, int], int] | None = None,
) -> BatchExpansionPlan:
    """Expand prompts × base seeds × variations into deterministic job specs."""
    normalized = normalize_prompts(
        prompts,
        max_prompts=max_prompts,
        max_prompt_length=max_prompt_length,
    )
    if variation_count > max_variations:
        raise BatchRequestInvalidError(
            f"variation_count may be at most {max_variations}.",
            field_issues={
                "variation_count": [
                    FieldIssue(
                        code=FieldIssueCode.VALUE_ABOVE_MAXIMUM,
                        message=f"variation_count may be at most {max_variations}.",
                        actual=variation_count,
                        maximum=max_variations,
                    )
                ]
            },
        )
    base_seeds = resolve_base_seeds(
        seed_mode=seed_mode,
        seed=seed,
        seed_start=seed_start,
        seed_end=seed_end,
        variation_count=variation_count,
        min_seed=min_seed,
        max_seed=max_seed,
        random_seed_factory=random_seed_factory,
    )
    stride = len(base_seeds)
    job_count = len(normalized) * stride * variation_count
    if job_count > max_jobs:
        raise BatchTooLargeError(
            f"This batch would create {job_count} jobs, which exceeds the maximum of {max_jobs}.",
            details={"job_count": job_count, "maximum": max_jobs},
            field_issues={
                "batch": [
                    FieldIssue(
                        code=FieldIssueCode.VALUE_ABOVE_MAXIMUM,
                        message=(
                            f"This batch would create {job_count} jobs "
                            f"(maximum {max_jobs}). Reduce prompts, seed range, or variations."
                        ),
                        actual=job_count,
                        maximum=max_jobs,
                    )
                ]
            },
        )

    items: list[BatchExpansionItem] = []
    warnings: list[str] = []
    seen_prompt_text: set[str] = set()
    for prompt_index, prompt in enumerate(normalized):
        if prompt in seen_prompt_text:
            warnings.append(
                f"Prompt {prompt_index} duplicates an earlier entry and reuses the same seeds."
            )
        seen_prompt_text.add(prompt)
        used_seeds: set[int] = set()
        for base in base_seeds:
            for variation_index in range(variation_count):
                actual_seed = base + variation_index * stride
                if actual_seed > max_seed:
                    raise BatchRequestInvalidError(
                        "Expanded seed exceeds the allowed maximum. "
                        "Reduce the sequential range or variation count.",
                        field_issues={
                            "seed": [
                                FieldIssue(
                                    code=FieldIssueCode.VALUE_ABOVE_MAXIMUM,
                                    message="Expanded seed exceeds the allowed maximum.",
                                    actual=actual_seed,
                                    maximum=max_seed,
                                )
                            ]
                        },
                    )
                if actual_seed in used_seeds:
                    raise BatchRequestInvalidError(
                        "Sequential seed expansion produced a duplicate seed for the same prompt.",
                        field_issues={
                            "seed": [
                                _issue(
                                    "Sequential seed expansion produced a duplicate seed.",
                                    actual=actual_seed,
                                )
                            ]
                        },
                    )
                used_seeds.add(actual_seed)
                items.append(
                    BatchExpansionItem(
                        index=len(items),
                        prompt_index=prompt_index,
                        variation_index=variation_index,
                        seed=actual_seed,
                        prompt=prompt,
                        output_name=build_output_name(
                            output_name,
                            prompt_index=prompt_index,
                            seed=actual_seed,
                            variation_index=variation_index,
                            max_length=max_output_name_length,
                        ),
                    )
                )
    return BatchExpansionPlan(
        items=tuple(items),
        seed_mode=(seed_mode or "").strip().lower(),
        base_seeds=base_seeds,
        variation_count=variation_count,
        warnings=tuple(warnings),
    )
